### PR TITLE
`grpc-kafka` mapping implementation

### LINKED ISSUE #184: Support gRPC Kafka mapping
```
---
name: test
bindings:
  grpc_kafka_proxy0:
    type: grpc-kafka
    kind: proxy
    exit: kafka_cache_client0
  kafka_cache_client0:
    type: kafka
    kind: cache_client
    exit: kafka_cache_server0
  kafka_cache_server0:
    type: kafka
    kind: cache_server
    exit: kafka_client0
  kafka_client0:
    type: kafka
    kind: client
    exit: tcp0
```

### PR DESCRIPTION
**Pull request by akrambek**
Fixes #184

### CODE REVIEW

File: incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaRouteConfig.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        return authorized.test(authorization);
+    }
+
+    boolean matches(
+        String16FW service,
+        String16FW method,
+        Array32FW<GrpcMetadataFW> metadataHeaders)
+    {
+        return when.isEmpty() || service != null && method != null &&
+            when.stream().anyMatch(m -> m.matches(service, method, metadataHeaders));
```
Comment (jfallows) [inline incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaRouteConfig.java:82]: Suggest renaming `metadataHeaders` to `metadata`.

File: incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaWithResolver.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        GrpcKafkaWithConfig with)
+    {
+        this.identityReplacer = identityReplacer;
+        this.with = with;
+        this.identityMatcher = IDENTITY_PATTERN.matcher("");
+    }
+
+    public GrpcKafkaWithResult resolve(
+        long authorization,
+        GrpcBeginExFW grpcBeginEx)
```
Comment (jfallows) [inline incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaWithResolver.java:64]: Just `beginEx` is okay here, we are in `grpc` context.

File: incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaWithResolver.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    {
+        String16FW topic = new String16FW(with.topic);
+        KafkaAckMode acks = with.acks;
+
+        final GrpcMetadataFW grpcIdempotencyKey = grpcBeginEx.metadata().matchFirst(m ->
+            GRPC_METADATA_NAME_IDEMPOTENCY_KEY.value().compareTo(m.name().value()) == 0);
+
+        OctetsFW correlationId = null;
+        if (grpcIdempotencyKey != null)
```
Comment (jfallows) [inline incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaWithResolver.java:73]: Just `idempotency` is okay here, we are in `grpc` context.

File: incubator/binding-grpc-kafka/src/test/java/io/aklivity/zilla/runtime/blinding/grpc/kafka/internal/stream/GrpcKafkaProxyIT.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    {
+        k3po.finish();
+    }
+
+    @Test
+    @Configuration("proxy.rpc.yaml")
+    @Specification({
+        "${grpc}/bidi.stream.rpc.rejected/client",
+        "${kafka}/bidi.stream.rpc.rejected/server"})
+    public void shouldRejectBiStreamRpc() throws Exception
```
Comment (jfallows) [inline incubator/binding-grpc-kafka/src/test/java/io/aklivity/zilla/runtime/blinding/grpc/kafka/internal/stream/GrpcKafkaProxyIT.java:138]: typo, should be `shouldRejectBidiStreamRpc`

File: incubator/command-log/src/main/java/io/aklivity/zilla/runtime/command/log/internal/LoggableStream.java [hunk 5]
```diff
@@ -963,7 +963,7 @@ private String asString(
             final OctetsFW headerName = header.name();
             final String formattedHeaderName = headerName.buffer().getStringWithoutLengthUtf8(
                 headerName.offset(), headerName.sizeof());
-            formatted = String.format("header[%s=[%d]]", formattedHeaderName, header.valueLen());
+            formatted = String.format("header[%s=%s]", formattedHeaderName, asString(header.value()));
```
Comment (jfallows) [inline incubator/command-log/src/main/java/io/aklivity/zilla/runtime/command/log/internal/LoggableStream.java:966]: We cannot assume that Kafka header values are strings.

File: incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/stream/GrpcKafkaProxyFactory.java [hunk 6]
```diff
@@ -0,0 +1,1723 @@
...
+
+    private final MutableDirectBuffer writeBuffer;
+    private final MutableDirectBuffer extBuffer;
+    private final BindingHandler streamFactory;
+    private final LongUnaryOperator supplyInitialId;
+    private final LongUnaryOperator supplyReplyId;
+    private final Signaler signaler;
+    private final int grpcTypeId;
+    private final int kafkaTypeId;
+
```
Comment (jfallows) [inline incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/stream/GrpcKafkaProxyFactory.java:113]: Please remove extra blank line.

**Review by jfallows — CHANGES_REQUESTED** [review]
For some of the negative tests, such as `*.rejected`, those are really testing `abort`.

In `grpc` binding specs, we use `rejected` to indicate the stream was not `accepted` because it was invalid in some way, so the meaning is a bit different.

Also, I'm not yet sure how we would convey via `kafka-grpc` that the `grpc` side did actually reject the `grpc` call, such that it would propagate back to `grpc-kafka` here and really reject the `grpc` stream.

File: incubator/command-log/src/main/java/io/aklivity/zilla/runtime/command/log/internal/LoggableStream.java (hunk 5 above)
Comment (akrambek) [reply incubator/command-log/src/main/java/io/aklivity/zilla/runtime/command/log/internal/LoggableStream.java:966]: Yeah, left over I was debugging

File: incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaConditionConfigAdapter.java [hunk 7]
```diff
@@ -105,22 +105,31 @@ public ConditionConfig adaptFromJson(
...
-                    textValue = JsonString.class.cast(v).getString();
+                case OBJECT:
+                    if (v.asJsonObject().containsKey(BASE64_NAME))
+                    {
+                        base64Value = v.asJsonObject().getString(BASE64_NAME);
+                    }
+                    else
+                    {
+                        throw new IllegalArgumentException("Unexpected metadata type: " + v.asJsonObject());
+                    }
```
Comment (jfallows) [inline incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaConditionConfigAdapter.java:122]: I think this validation code can never execute due to schema checks, right?

File: incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaConditionConfigAdapter.java [hunk 8]
```diff
@@ -105,22 +105,31 @@ public ConditionConfig adaptFromJson(
...
+                    break;
+                case STRING:
+                    textValue = ((JsonString) v).getString();
                     base64Value = encoder64.encodeToString(textValue.getBytes());
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unexpected type: " + valueType);
```
Comment (jfallows) [inline incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaConditionConfigAdapter.java:129]: This default case can never execute either, agree?

File: incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaConditionConfigAdapter.java [hunk 9]
```diff
@@ -105,22 +105,31 @@ public ConditionConfig adaptFromJson(
...
                 }
 
-                GrpcKafkaMetadataValue metadataValue = new GrpcKafkaMetadataValue(new String16FW(textValue),
+                GrpcKafkaMetadataValue metadataValue =  new GrpcKafkaMetadataValue(new String16FW(textValue),
```
Comment (jfallows) [inline incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaConditionConfigAdapter.java:132]: Please remove extra space after `=`, this is probably a checkstyle violation too. 🙂

File: incubator/binding-grpc-kafka/src/test/java/io/aklivity/zilla/runtime/blinding/grpc/kafka/internal/stream/GrpcKafkaProxyIT.java [hunk 10]
```diff
@@ -75,7 +75,7 @@ public void shouldRejectUnaryRpc() throws Exception
     @Specification({
         "${grpc}/unary.rpc.sent.write.abort/client",
         "${kafka}/unary.rpc.sent.write.abort/server"})
-    public void shouldHandleUnaryRrpSentWriteAbort() throws Exception
+    public void shouldNotProduceMessageOnUnaryRrcSentWriteAbort() throws Exception
```
Comment (jfallows) [inline incubator/binding-grpc-kafka/src/test/java/io/aklivity/zilla/runtime/blinding/grpc/kafka/internal/stream/GrpcKafkaProxyIT.java:78]: Typo: `Rrp` -> `Rpc`

**Review by jfallows — CHANGES_REQUESTED** [review]

File: incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaConditionConfigAdapter.java (hunk 7 above)
Comment (akrambek) [reply incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaConditionConfigAdapter.java:122]: Yeah, schema validation

**Review by jfallows — APPROVED** [review]

**Pull request merged**